rdt-app/src/main/java/io/ona/rdt_app/interactor/PatientRegisterFragmentInteractor.java

```diff
@@ -167,4 +176,40 @@ public Patient getPatientForRDT(JSONObject jsonFormObject) throws JSONException
         return rdtPatient;
     }
 
+    public synchronized void launchForm(Activity activity, String formName, Patient patient) throws JSONException {
+        try {
+            JSONObject formJsonObject = formUtils.getFormJsonObject(formName, activity);
+            if (patient != null) {
+                FormLaunchArgs args = new FormLaunchArgs().withActivity(activity)
+                        .withPatient(patient)
+                        .withFormJsonObj(formJsonObject);
+                formUtils.getNextUniqueId(args, this);
+            } else {
+                formUtils.prePopulateFormFields(formJsonObject, patient, "", 2); // todo: see if numfields is correct here
+                formUtils.startJsonForm(formJsonObject, activity, REQUEST_CODE_GET_JSON);
+            }
+        } catch (JsonFormMissingStepCountException e) {
+            Log.e(TAG, e.getStackTrace().toString());
+        }
+    }
+
+    @Override
+    public synchronized void onUniqueIdFetched(FormLaunchArgs args, UniqueId uniqueId) {
+        try {
+            Activity activity = args.getActivity();
+            String id = uniqueId == null ? "" : uniqueId.getOpenmrsId().replace("-", "");
+            if (id.isEmpty()) {
+                showToast(activity, activity.getString(R.string.unique_id_fetch_error_msg));
+            } else {
+                RDTApplication.getInstance().getContext().getUniqueIdRepository().close(id);
```

Comment:
You should close the Id when saving the event and not here when generating the id. If the form is not saved then the Ids will not have been used but will be marked as closed

Reply:
This would require a non-trivial refactor to pass around the rdt id. 

Or alternatively need to traverse the form json object when saving to extract the rdt id and close it.

I don't think this instance (where saving doesn't happen) is that common to require this. Is there an alternative way you thought to address closing the id during save?

Reply:
Yes the Id is saved as an Obs `Constants.Form.LBL_RDT_ID`.

So extract this obs and retrieve the value when saving and close the ID. 

You can use `org.smartregister.domain.db.Event#findObs` to get the Obs

Reply:
Cool, done.